### ISSUE TITLE
DBAL-1245: Unexpected behavior when getting schema updates from ORM

### ISSUE THREAD

**Issue by doctrinebot** [doctrine/dbal#1195]
Jira issue originally created by user tgeorges:

Hi all, I found a problem using DBAL and schema update : 

My setup : 

DBAL 2.5.1, Oracle OCI

I ask for schema updates using command 

> php app/console doctrine:schema:update --dump-sql

Result : 

```
ALTER INDEX idx*ab074de1ef5927f RENAME TO IDX*FFF42BF5727ACA70;
ALTER INDEX idx*218e7204ef5927f RENAME TO IDX*6E75D7727ACA70;
ALTER INDEX idx*85c03c85ee8c0707 RENAME TO IDX*1AB76DA1B52C0544;
ALTER INDEX idx*523305bac5653160 RENAME TO IDX*D6AFDC364E833AFF;
ALTER INDEX idx*523305babb38df8d RENAME TO IDX*D6AFDC36ED1C4E5;
ALTER INDEX idx*523305baee8c0707 RENAME TO IDX*D6AFDC36B52C0544;
ALTER INDEX idx*ddcece30d8abdcf4 RENAME TO IDX*22ABD2A38661593;
ALTER INDEX idx*6743b99ac5653160 RENAME TO IDX*5C4ACA974E833AFF;
ALTER INDEX idx*6743b99a691f1cfc RENAME TO IDX*5C4ACA974C5B503A;
ALTER INDEX idx*fc1af60cf41679cd RENAME TO IDX*FC1AF60C15A17E77;
ALTER INDEX idx*4c268f82f41679cd RENAME TO IDX*4C268F82C7C42212;
ALTER INDEX idx*4c268f82a6e00f45 RENAME TO IDX*4C268F82DA6F574A;
ALTER INDEX idx*7ca8bc6df41679cd RENAME TO IDX*7CA8BC6D15A17E77;
...
```

After investigating, Doctrine\ORM\Tools\SchemaTool asks the "from" Schema to OracleSchemaManager and the "to" schema is build from metadata information.
The from schema list each tables and their content : columns, indexex and constraints from Oracle Database. ListTables result is well formed.

After that, SchemaManager build a Doctrine\DBAL\Schema\Table object (in listTableDetails) to represent each table.
Table construct index associated to constraint (internally considered as implicit index).

This index is then interpreted as an existing database index by Comparator : 
it generates a SchemaDiff object that contains SQL statements "Rename" or "Drop/Create" (depends of DBAL versions) on these implicit indexes that does not exist in database.

To sum up

```
- SchemaTool ask schema "from" to SchemaManager
- SchemaManager read schema from database (OK)
- Table generate non existing implicit index (OK)
- Comparator generate diff and consider these indexes as existing (NOK) 
```

Today I fix it by appending a contructor argument to Table in order to not generate implicit indexes.
I think this can also be fixed by opening the concept of implicit index to public namespace and implement its usage in Comparator and other tools because this information can be very usefull in other ORM tools.

Do you think there is another approach for my problem ?

Thanks for reading,

Tony

**Comment by tgeorges** [comment]
Please contact me if you need any help to reproduce this issue

**Comment by deeky666** [comment]
@tgeorges can you provide a failing test case for DBAL? Or can you at least provide an example of how to exactly reproduce this issue? Looks like it has to do with the index renaming feature introduced in 2.5.

**Comment by tgeorges** [comment]
@deeky666 this problem is old for me since I solve it with a patch. As I remember, when I update my schema to reflect PHP classes, dbal tray to rename indexes instead of creating them. 

My colleague @Charlewic is working on this project now, so I tell him to help you in reproducing this issue

**Comment by deeky666** [comment]
@tgeorges okay thank you. Waiting for feedback then :)

**Comment by zeroedin-bill** [comment]
@deeky666 I think I can repro this.

**Comment by deeky666** [comment]
Okay assigning this to you then for now.

**Comment by Charlewic** [comment]
Hello,

Have you some fresh news on this issue ?
Need for more feedback ?

Thank for your answers

**Comment by digilist** [comment]
I ran into the same issue. The problem occurs for me with MySQL, when I have a foreign key where the index is a combined index. Consider this table structure:


	CREATE TABLE `ref` (
		`id` int (11) NOT NULL AUTO_INCREMENT,
		PRIMARY KEY (`id`)
	) ENGINE=InnoDB DEFAULT CHARSET=utf8;

	CREATE TABLE `test` (
	  `id` int(11) NOT NULL AUTO_INCREMENT,
	  `ref_id` int(11) DEFAULT NULL,
	  `other_column` int(11) NOT NULL,
	  PRIMARY KEY (`id`),
	  UNIQUE KEY `foo` (`ref_id`,`other_column`)
	) ENGINE=InnoDB  DEFAULT CHARSET=utf8;

	ALTER TABLE `test` ADD CONSTRAINT `fk_ref` FOREIGN KEY (`ref_id`) REFERENCES `ref` (`id`);

So, there is a index for the foreign key which is perfectly fine (at least in MySQL, don't know about other database engines). However, Doctrine thinks that there is no index because the `isFullfilledBy()` method considers the number of columns in an index. The foreign key consists only of one column while the `foo` key consists of two columns. Therefore, the implicit index is not fulfilled by the existing index and Doctrine adds an index for a single column (`ref_id`). Later during the diff, Doctrine detects correctly that there should not be such an index and tries to drop it. However, the index does not exist in the database.

(Disclaimer: I didn't test the SQL Statement above, but as far as I researched the problem, it leads to this issue)
